File: javascript/medium/subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const res = new Array(), temp = new Array();
    generateSubsets(0, nums, temp, res);
    return res;
};

var generateSubsets = function(index, nums, temp, res) {
    res.push(temp.slice(0));
    for(let i = index; i < nums.length; i++) {
        temp.push(nums[i]);
        generateSubsets(i+1, nums, temp, res);
        temp.pop();
    }
};

File: ruby/medium/subsets.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def subsets(nums)
    res, temp = Array.new, Array.new
    generate(0, nums, temp, res)
    return res
end

def generate(index, nums, temp, res)
    res.push(temp.dup)
    for i in index..nums.length-1
        temp.push(nums[i])
        generate(i+1, nums, temp, res)
        temp.pop
    end
end
